README.rst
httpsig
=======

.. image:: https://travis-ci.org/ahknight/httpsig.svg?branch=master
    :target: https://travis-ci.org/ahknight/httpsig

.. image:: https://travis-ci.org/ahknight/httpsig.svg?branch=develop
    :target: https://travis-ci.org/ahknight/httpsig

Sign HTTP requests with secure signatures according to the IETF HTTP Signatures specification from draft 00 to 07.  This is a fork of the original module_ to fully support both RSA and HMAC schemes as well as unit test both schemes to prove they work.  It's being used in production and is actively-developed.

See the original project_, original Python module_, original spec_, and `current IETF draft`_ for more details on the signing scheme.

.. _project: https://github.com/joyent/node-http-signature
.. _module: https://github.com/zzsnzmn/py-http-signature
.. _spec: https://github.com/joyent/node-http-signature/blob/master/http_signing.md
.. _`current IETF draft`: https://datatracker.ietf.org/doc/draft-cavage-http-signatures/

Requirements
------------

* Python 2.7, 3.3, 3.4, 3.5, 3.6
* PyCrypto_

Optional:

* requests_

.. _PyCrypto: https://pypi.python.org/pypi/pycrypto
.. _requests: https://pypi.python.org/pypi/requests

Usage
-----

Real documentation is forthcoming, but for now this should get you started.

For simple raw signing:

.. code:: python

    import httpsig

    secret = open('rsa_private.pem', 'rb').read()

    sig_maker = httpsig.Signer(secret=secret, algorithm='rsa-sha256')
    sig_maker.sign('hello world!')

For general use with web frameworks:

.. code:: python

    import httpsig

    key_id = 'some key ID'
    secret = 'some big secret'

    hs = httpsig.HeaderSigner(key_id, secret, algorithm='hmac-sha256', headers=['(request-target)', 'host', 'date'])
    signed_headers_dict = hs.sign({'Date': 'Tue, 01 Jan 2014 01:01:01 GMT', 'Host': 'example.com'}, method='GET', path='/api/1/object/1')

For use with requests:

.. code:: python

    import json
    import requests
    from httpsig.requests_auth import HTTPSignatureAuth

    secret = open('rsa_private.pem', 'rb').read()

    auth = HTTPSignatureAuth(key_id='Test', secret=secret)
    z = requests.get('https://api.example.com/path/to/endpoint',
                             auth=auth, headers={'X-Api-Version': '~6.5'})

Class initialization parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Note that keys and secrets should be bytes objects.  At attempt will be made to convert them, but if that fails then exceptions will be thrown.

.. code:: python

    httpsig.Signer(secret, algorithm='rsa-sha256')

:secret: in the case of an RSA signature, is a string containing private RSA pem. In the case of HMAC, it is a secret password.
:algorithm: is one of the six allowed hash-sign algorithm combinations: ``rsa-sha1``, ``rsa-sha256``, ``rsa-sha512``, ``hmac-sha1``, ``hmac-sha256``, ``hmac-sha512``.


.. code:: python

    httpsig.requests_auth.HTTPSignatureAuth(key_id, secret, algorithm='rsa-sha256', headers=None, httpsig_version=None)

:key_id: is the label by which the server system knows your RSA signature or password.
:headers: is the list of HTTP headers that are concatenated and used as signing objects. By default it is the specification's minimum, the ``Date`` HTTP header.
:httpsig_version: is the IEFT version. By default it is ``draft-07`` and allowed: ``draft-00`` to ``draft-07``.
:secret:  as above.
:algorithm: as above.

Tests
-----

To run tests::

    python setup.py test

or::

    tox

License
-------

Both this module and the original module_ are licensed under the MIT license.


CHANGELOG.rst
httpsig Changes
---------------

1.1.3 (2017-July)
-------------------

* Support IEFT ``draft-00`` to ``draft-07``
* Python version update: support ``py3.5``, ``py3.6`` and remove ``py3.2`` (due to coverage not supported)
* Python re-format by using linter.
* Complete unittest and enhance coverage to over 90%.
* Bugfix for python3 string encode and decode.


1.1.2 (2015-Feb-11)
-------------------

* HMAC verification is now constant-time.

1.1.1 (2015-Feb-11)
-------------------

* (pulled)

1.1.0 (2014-Jul-24)
-------------------

* Changed "(request-line)" to "(request-target)" to comply with Draft 3.

1.0.3 (2014-Jul-09)
-------------------

* Unified the default signing algo under one setting. Setting httpsig.sign.DEFAULT_SIGN_ALGORITHM changes it for all future instances.
* Handle invalid params a little better.

1.0.2 (2014-Jul-02)
-------------------

* Ensure we treat headers as ASCII strings.
* Handle a case in the authorization header where there's garbage (non-keypairs) after the method name.

1.0.1 (2014-Jul-02)
~~~~~~~~~~~~~~~~~~~

* Python 3 support (2.7 + 3.2-3.4)
* Updated tox and Travis CI configs to test the supported Python versions.
* Updated README.

1.0.0 (2014-Jul-01)
~~~~~~~~~~~~~~~~~~~
* Written against http://tools.ietf.org/html/draft-cavage-http-signatures-02
* Added "setup.py test" and tox support.
* Added sign/verify unit tests for all currently-supported algorithms.
* HeaderSigner and HeaderVerifier now share the same message-building logic.
* The HTTP method in the message is now properly lower-case.
* Resolved unit test failures.
* Updated Verifier and HeaderVerifier to handle verifying both RSA and HMAC sigs.
* Updated versioneer.
* Updated contact/author info.
* Removed stray keypair in test dir.
* Removed SSH agent support.
* Removed suport for reading keyfiles from disk as this is a huge security hole if this is used in a server framework like drf-httpsig.

1.0b1 (2014-Jun-23)
~~~~~~~~~~~~~~~~~~~~~~
* Removed HTTP version from request-line, per spec (breaks backwards compatability).
* Removed auto-generation of missing Date header (ensures client compatability).


http-signature (previous)
-------------------------

0.2.0 (unreleased)
~~~~~~~~~~~~~~~~~~

* Update to newer spec (incompatible with prior version).
* Handle `request-line` meta-header.
* Allow secret to be a PEM encoded string.
* Add test cases from spec.

0.1.4 (2012-10-03)
~~~~~~~~~~~~~~~~~~

* Account for ssh now being re-merged into paramiko: either package is acceptable (but paramiko should ideally be >= 1.8.0)

0.1.3 (2012-10-02)
~~~~~~~~~~~~~~~~~~

* Stop enabling `allow_agent` by default
* Stop requiring `ssh` package by default -- it is imported only when `allow_agent=True`
* Changed logic around ssh-agent: if one key is available, don't bother with any other authentication method
* Changed logic around key file usage: if decryption fails, prompt for password
* Bug fix: ssh-agent resulted in a nonsensical error if it found no correct keys (thanks, petervolpe)
* Introduce versioneer.py


tests/test_sign.py
#!/usr/bin/env python
import os
import unittest
import six

import httpsig.sign as sign
from httpsig.utils import parse_authorization_header, HttpSigException


class TestSign(unittest.TestCase):

    def setUp(self):
        self.key_path = os.path.join(os.path.dirname(__file__), 'rsa_private.pem')
        with open(self.key_path, 'rb') as f:
            self.key = f.read()

    def test_default(self):
        hs = sign.HeaderSigner(key_id='Test', secret=self.key, algorithm='rsa-sha256')
        unsigned = {
            'Date': 'Thu, 05 Jan 2012 21:31:40 GMT'
        }
        signed = hs.sign(unsigned)
        self.assertIn('Date', signed)
        self.assertEqual(unsigned['Date'], signed['Date'])
        self.assertIn('Authorization', signed)
        auth = parse_authorization_header(signed['authorization'])
        params = auth[1]
        self.assertIn('keyId', params)
        self.assertIn('algorithm', params)
        self.assertIn('signature', params)
        self.assertEqual(params['keyId'], 'Test')
        self.assertEqual(params['algorithm'], 'rsa-sha256')
        self.assertEqual(params['signature'], 'ATp0r26dbMIxOopqw0OfABDT7CKMIoENumuruOtarj8n/97Q3htHFYpH8yOSQk3Z5zh8UxUym6FYTb5+A0Nz3NRsXJibnYi7brE/4tx5But9kkFGzG+xpUmimN4c3TMN7OFH//+r8hBf7BT9/GmHDUVZT2JzWGLZES2xDOUuMtA=')

    def test_all(self):
        hs = sign.HeaderSigner(key_id='Test', secret=self.key, algorithm='rsa-sha256', headers=[
            '(request-target)',
            'host',
            'date',
            'content-type',
            'content-md5',
            'content-length'
        ])
        unsigned = {
            'Host': 'example.com',
            'Date': 'Thu, 05 Jan 2012 21:31:40 GMT',
            'Content-Type': 'application/json',
            'Content-MD5': 'Sd/dVLAcvNLSq16eXua5uQ==',
            'Content-Length': '18',
        }
        signed = hs.sign(unsigned, method='POST', path='/foo?param=value&pet=dog')

        self.assertIn('Date', signed)
        self.assertEqual(unsigned['Date'], signed['Date'])
        self.assertIn('Authorization', signed)
        auth = parse_authorization_header(signed['authorization'])
        params = auth[1]
        self.assertIn('keyId', params)
        self.assertIn('algorithm', params)
        self.assertIn('signature', params)
        self.assertEqual(params['keyId'], 'Test')
        self.assertEqual(params['algorithm'], 'rsa-sha256')
        self.assertEqual(params['headers'], '(request-target) host date content-type content-md5 content-length')
        self.assertEqual(params['signature'], 'G8/Uh6BBDaqldRi3VfFfklHSFoq8CMt5NUZiepq0q66e+fS3Up3BmXn0NbUnr3L1WgAAZGplifRAJqp2LgeZ5gXNk6UX9zV3hw5BERLWscWXlwX/dvHQES27lGRCvyFv3djHP6Plfd5mhPWRkmjnvqeOOSS0lZJYFYHJz994s6w=')

    def test_PASS_verify_headers_by_draft_version(self):
        testcases = {
            'draft-00': 'request-line',
            'draft-01': 'request-line',
            'draft-02': '(request-line)',
            'draft-03': '(request-target)',
            'draft-04': '(request-target)',
            'draft-05': '(request-target)',
            'draft-06': '(request-target)',
            'draft-07': '(request-target)',
            None: '(request-target)'
        }
        for httpsig_version, header_req in testcases.items():
            hs = sign.HeaderSigner(key_id='Test', secret=self.key, httpsig_version=httpsig_version, headers=[
                header_req,
                'host',
                'date',
                'content-type',
                'content-md5',
                'content-length'
            ])

    def test_FAIL_verify_headers_by_draft_version(self):
        testcases = [
            ('draft-00', '(request-line)'),
            ('draft-00', '(request-target)'),
            ('draft-01', '(request-line)'),
            ('draft-01', '(request-target)'),
            ('draft-02', 'request-line'),
            ('draft-02', '(request-target)'),
            ('draft-03', 'request-line'),
            ('draft-03', '(request-line)'),
            ('draft-04', 'request-line'),
            ('draft-04', '(request-line)'),
            ('draft-05', 'request-line'),
            ('draft-05', '(request-line)'),
            ('draft-06', 'request-line'),
            ('draft-06', '(request-line)'),
            ('draft-07', 'request-line'),
            ('draft-07', '(request-line)'),
            (None, 'request-line'),
            (None, '(request-line)'),
        ]
        for httpsig_version, header_req in testcases:
            try:
                sign.HeaderSigner(key_id='Test', secret=self.key, httpsig_version=httpsig_version, headers=[
                    header_req,
                    'host',
                    'date',
                    'content-type',
                    'content-md5',
                    'content-length'
                ])
                self.fail('Should raise KeyError in (%s, %s)' % (httpsig_version, header_req))
            except KeyError:
                pass

    def test_signer_init(self):

        signature = 'ksTqAWZHgRktH3uKbXydPNEMqpjqALFiJVJXeCKj8zA='
        secret = 'i am a secret'
        data = 'i am some data'
        algorithm = 'hmac-sha256'

        # test good case: default string type
        signer = sign.Signer(secret, algorithm=algorithm)
        self.assertEqual(signer.sign(data), signature)
        self.assertEqual(signer.algorithm, algorithm)

        # test good case: unicode/byte string type for secret
        signer = sign.Signer(secret.encode() if six.PY3 else secret.decode(), algorithm=algorithm)
        self.assertEqual(signer.sign(data), signature)

        # test good case: unicode/byte string type for data
        signer = sign.Signer(secret, algorithm=algorithm)
        self.assertEqual(signer.sign(data.encode() if six.PY3 else data.decode()), signature)

        # test invalid hash algorithm
        weird_algorithm = 'rsa-sha257'
        original_algorithms = sign.ALGORITHMS
        sign.ALGORITHMS = frozenset([weird_algorithm])
        try:
            signer = sign.Signer(secret, algorithm=weird_algorithm)
            self.fail('No exception was raised')
        except Exception as ex:
            self.assertIsInstance(ex, HttpSigException)
        finally:
            sign.ALGORITHMS = original_algorithms

        # test invalid sign algorithm
        weird_algorithm = 'xxx-sha256'
        original_algorithms = sign.ALGORITHMS
        sign.ALGORITHMS = frozenset([weird_algorithm])
        try:
            signer = sign.Signer(secret, algorithm=weird_algorithm)
            signer.sign('nothing')
            self.fail('No exception was raised')
        except Exception as ex:
            self.assertIsInstance(ex, SystemError)
        finally:
            sign.ALGORITHMS = original_algorithms


tests/test_utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import unittest
import six
import httpsig.utils


class TestUtils(unittest.TestCase):

    def test_ct_bytes_compare(self):
        if six.PY3:
            test_cases = [
                (b"\xa5Q\x7f\x92q\x8b+\xd5\x83x'^", b"\xa5Q\x7f\x92q\x8b+\xd5\x83x'^", True),
                ('123', '123', True),
                (b'123', '123', True),
                (b'123', b'123', True),
                (b'\xe5\xb0\x8d', '對', True),
                (b'123456', b'123', False)
            ]
        elif six.PY2:
            test_cases = [
                ("\xa5Q\x7f\x92q\x8b+\xd5\x83x'^", "\xa5Q\x7f\x92q\x8b+\xd5\x83x'^", True),
                ('123', '123', True),
                ('123', u'123', True),
                (u'123', u'123', True),
                ('\xe5\xb0\x8d', u'對', True),
                ('123456', '123', False)
            ]

        for case in test_cases:
            self.assertIs(httpsig.utils.ct_bytes_compare(case[0], case[1]), case[2])

    def test_generate_message(self):
        HOST = "example.org"
        METHOD = "POST"
        PATH = '/foo'
        HTTP_VERSION = 'HTTP/1.1'
        HEADERS = {
            'Host': 'example.org',
            'Date': 'Tue, 07 Jun 2014 20:51:35 GMT',
            'Content-Type': 'application/json',
            'Digest': 'SHA-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=',
            'Content-Length': 18
        }
        good_test_cases = {
            # from draft-01: 3.1.2.  RSA Example
            'request-line host date digest content-length': b'POST /foo HTTP/1.1\nhost: example.org\ndate: Tue, 07 Jun 2014 20:51:35 GMT\ndigest: SHA-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=\ncontent-length: 18',
            # draft-02
            '(request-line) host date digest content-length': b'(request-line): post /foo\nhost: example.org\ndate: Tue, 07 Jun 2014 20:51:35 GMT\ndigest: SHA-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=\ncontent-length: 18',
            # draft-03
            '(request-target) host date digest content-length': b'(request-target): post /foo\nhost: example.org\ndate: Tue, 07 Jun 2014 20:51:35 GMT\ndigest: SHA-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=\ncontent-length: 18'
        }

        for required_headers, result in good_test_cases.items():
            assert httpsig.utils.generate_message(
                required_headers=required_headers.split(),
                headers=HEADERS,
                host=HOST,
                method=METHOD,
                path=PATH,
                http_version=HTTP_VERSION) == result, 'header: %s\nexpect: %s\n' % (required_headers, result)

        exception_test_cases = [
            {'required_headers': ['(request-target)'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '', 'http_version': ''},
            {'required_headers': ['(request-target)'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': 'GET', 'path': '', 'http_version': ''},
            {'required_headers': ['(request-target)'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '/', 'http_version': ''},
            {'required_headers': ['(request-line)'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': ''},
            {'required_headers': ['(request-line)'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': 'GET', 'path': '', 'http_version': ''},
            {'required_headers': ['(request-line)'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '/', 'http_version': ''},
            {'required_headers': ['request-line'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '', 'http_version': ''},
            {'required_headers': ['request-line'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': 'GET', 'path': '', 'http_version': ''},
            {'required_headers': ['request-line'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '/', 'http_version': ''},
            {'required_headers': ['request-line'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '', 'http_version': 'HTTP/1.1'},
            {'required_headers': ['request-line'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': 'GET', 'path': '', 'http_version': 'HTTP/1.1'},
            {'required_headers': ['request-line'], 'host': HOST, 'headers': HEADERS, 'exception': KeyError, 'method': '', 'path': '/', 'http_version': 'HTTP/1.1'},
            {'required_headers': [], 'host': HOST, 'headers': [], 'exception': KeyError, 'method': '', 'path': '', 'http_version': ''},
            {'required_headers': ['haha'], 'host': HOST, 'headers': [], 'exception': KeyError, 'method': '', 'path': '', 'http_version': ''},
            {'required_headers': ['host'], 'host': '', 'headers': [], 'exception': KeyError, 'method': '', 'path': '', 'http_version': ''}
        ]
        for case in exception_test_cases:
            try:
                httpsig.utils.generate_message(
                    required_headers=case['required_headers'],
                    headers=case['headers'],
                    host=case['host'],
                    method=case['method'],
                    path=case['path'],
                    http_version=case['http_version'])
                self.fail('No exception was raised')
            except Exception as ex:
                self.assertIsInstance(ex, case['exception'], str(case))

    def test_parse_authorization_header(self):
        if six.PY3:
            good_test_cases = [
                (
                    'Signature keyId="Test",algorithm="rsa-sha512",headers="(request-target) host date content-type content-md5 content-length"',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'algorithm': 'rsa-sha512',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                ),
                (
                    b'Signature keyId="Test",algorithm="rsa-sha512",headers="(request-target) host date content-type content-md5 content-length"',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'algorithm': 'rsa-sha512',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                ),
                (
                    'Signature keyId="Test",algorithm=,headers="(request-target) host date content-type content-md5 content-length",=123',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                ),
                (
                    'Signature keyId="Test", algorithm="rsa-sha512" ,headers="(request-target) host date content-type content-md5 content-length"',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'algorithm': 'rsa-sha512',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                )
            ]
        elif six.PY2:
            good_test_cases = [
                (
                    'Signature keyId="Test",algorithm="rsa-sha512",headers="(request-target) host date content-type content-md5 content-length"',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'algorithm': 'rsa-sha512',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                ),
                (
                    u'Signature keyId="Test",algorithm="rsa-sha512",headers="(request-target) host date content-type content-md5 content-length"',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'algorithm': 'rsa-sha512',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                ),
                (
                    'Signature keyId="Test",algorithm=,headers="(request-target) host date content-type content-md5 content-length",=123',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                ),
                (
                    'Signature keyId="Test", algorithm="rsa-sha512" ,headers="(request-target) host date content-type content-md5 content-length"',
                    ('Signature', httpsig.utils.CaseInsensitiveDict({
                        'keyId': 'Test',
                        'algorithm': 'rsa-sha512',
                        'headers': '(request-target) host date content-type content-md5 content-length'}))
                )
            ]

        for case in good_test_cases:
            self.assertEqual(httpsig.utils.parse_authorization_header(case[0]), case[1])

        exception_test_cases = [
            ('Signature-keyId="Test",algorithm="rsa-sha512"', ValueError),
            ('', ValueError),
            (123, AttributeError)
        ]
        for case in exception_test_cases:
            try:
                httpsig.utils.parse_authorization_header(case[0])
                self.fail('No exception was raised.')
            except Exception as ex:
                self.assertIsInstance(ex, case[1])

    def test_get_fingerprint(self):
        with open(os.path.join(os.path.dirname(__file__), 'rsa_public.pem'), 'r') as k:
            key = k.read()
        fingerprint = httpsig.utils.get_fingerprint(key)
        self.assertEqual(fingerprint, "73:61:a2:21:67:e0:df:be:7e:4b:93:1e:15:98:a5:b7")

        test_case = [('ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDFuqGVUQjw8KtwY+JIZx4uGhy2ap4QLjtUOaH/o8vxUeAk7P5Olhxzr2FBnCUjS6iZmuzZzviXI3NhyR2ic661hFlXxkJaEa6DruRakZ6P+uMFPmvE+RsOp0ppcW2uGO5Y8C0OqEMI4NT2E4/LIzM7kmspF7cvJajUa9UQ6ZpKG/YfZpOs6xug8uT+1GCiPC+w/GX2UWtj2kmTUUJZWddSev9kHUDbPl6GwLMmnJ3UB9C7rdNlhupArJAsL+7eAXR9DcV5Fo7kDtFYiZllwRAWVghVjeGyEkSOEEbVCtI+l+V2cFYlta1mPXJQsshKvzYQ25IjnDBnXGg/HpwCppWT httpsig', '99:18:ae:30:be:c4:12:ce:b5:17:2b:56:ee:a9:ab:23')]

        for case in test_case:
            self.assertEqual(httpsig.utils.get_fingerprint(case[0]), case[1])


httpsig/sign.py
import base64
import six

from Crypto.Hash import HMAC
from Crypto.PublicKey import RSA
from Crypto.Signature import PKCS1_v1_5

from .utils import ALGORITHMS, HASHES, HttpSigException, build_signature_template, CaseInsensitiveDict, generate_message


DEFAULT_SIGN_ALGORITHM = 'hmac-sha256'
DEFAULT_HTTPSIG_VERSION = 'draft-07'
SUPPORTED_HTTPSIG_VERSION = ['draft-00', 'draft-01', 'draft-02', 'draft-03', 'draft-04', 'draft-05', 'draft-06', 'draft-07']


class Signer(object):
    """
    When using an RSA algo, the secret is a PEM-encoded private key.
    When using an HMAC algo, the secret is the HMAC signing secret.

    Password-protected keyfiles are not supported.
    """
    def __init__(self, secret, algorithm=DEFAULT_SIGN_ALGORITHM):
        assert algorithm in ALGORITHMS, 'Unknown algorithm'
        if isinstance(secret, six.string_types):
            secret = secret.encode('ascii')

        self._rsa = None
        self._hash = None
        self.sign_algorithm, self.hash_algorithm = algorithm.split('-')

        if self.sign_algorithm == 'rsa':
            try:
                rsa_key = RSA.importKey(secret)
                self._rsa = PKCS1_v1_5.new(rsa_key)
                self._hash = HASHES[self.hash_algorithm]
            except ValueError:
                raise HttpSigException('Invalid key.')

        elif self.sign_algorithm == 'hmac':
            self._hash = HMAC.new(secret, digestmod=HASHES[self.hash_algorithm])

    @property
    def algorithm(self):
        return '%s-%s' % (self.sign_algorithm, self.hash_algorithm)

    def __sign_rsa(self, data):
        _hash = self._hash.new()
        _hash.update(data)
        return self._rsa.sign(_hash)

    def __sign_hmac(self, data):
        hmac = self._hash.copy()
        hmac.update(data)
        return hmac.digest()

    def _sign(self, data):
        '''return raw signed string'''
        signed = None
        if self._rsa:
            signed = self.__sign_rsa(data)
        elif self._hash:
            signed = self.__sign_hmac(data)
        if not signed:
            raise SystemError('No valid encryptor found.')
        return signed

    def sign(self, data):
        if isinstance(data, six.string_types):
            data = data.encode('ascii')

        return base64.b64encode(self._sign(data)).decode('ascii')

class HeaderSigner(Signer):
    '''
    Generic object that will sign headers as a dictionary using the http-signature scheme.
    https://github.com/joyent/node-http-signature/blob/master/http_signing.md

    :arg key_id:    the mandatory label indicating to the server which secret to use
    :arg secret:    a PEM-encoded RSA private key or an HMAC secret (must match the algorithm)
    :arg algorithm: one of the six specified algorithms
    :arg headers:   a list of http headers to be included in the signing string, defaulting to ['date'].
    '''
    def __init__(self, key_id, secret, algorithm=DEFAULT_SIGN_ALGORITHM, headers=None, httpsig_version=DEFAULT_HTTPSIG_VERSION):
        super(HeaderSigner, self).__init__(secret=secret, algorithm=algorithm)
        self.headers = headers or ['date']
        self.signature_template = build_signature_template(key_id, algorithm, headers)
        if httpsig_version in SUPPORTED_HTTPSIG_VERSION:
            self.httpsig_version = httpsig_version
        else:
            self.httpsig_version = DEFAULT_HTTPSIG_VERSION
        self._verify_headers_by_draft_version()

    def sign(self, headers, host=None, method=None, path=None):
        """
        Add Signature Authorization header to case-insensitive header dict.

        headers is a case-insensitive dict of mutable headers.
        host is a override for the 'host' header (defaults to value in headers).
        method is the HTTP method (required when using '(request-target)').
        path is the HTTP path (required when using '(request-target)').
        """
        headers = CaseInsensitiveDict(headers)
        signable = generate_message(self.headers, headers, host, method, path)

        signature = super(HeaderSigner, self).sign(signable)
        headers['authorization'] = self.signature_template % signature

        return headers

    def _verify_headers_by_draft_version(self):
        HEADER_MAPING = {
            'request-line': ['draft-00', 'draft-01'],
            '(request-line)': ['draft-02'],
            '(request-target)': ['draft-03', 'draft-04', 'draft-05', 'draft-06', 'draft-07']
        }

        for header, drafts in HEADER_MAPING.items():
            if header in self.headers:
                if self.httpsig_version not in drafts:
                    raise KeyError('%s is not supported by %s' % (header, self.httpsig_version))


httpsig/utils.py
import re
import hashlib
import base64
from functools import reduce
import six

try:
    # Python 3
    from urllib.request import parse_http_list
except ImportError:  # pragma: no cover
    # Python 2
    from urllib2 import parse_http_list

from Crypto.Hash import SHA, SHA256, SHA512

ALGORITHMS = frozenset(['rsa-sha1', 'rsa-sha256', 'rsa-sha512', 'hmac-sha1', 'hmac-sha256', 'hmac-sha512'])
HASHES = {'sha1': SHA,
          'sha256': SHA256,
          'sha512': SHA512}


class HttpSigException(Exception):
    pass


def ct_bytes_compare(byte_l, byte_r):
    """
    Constant-time string compare.
    http://codahale.com/a-lesson-in-timing-attacks/
    """
    if not isinstance(byte_l, six.binary_type):
        byte_l = byte_l.encode('utf8')
    if not isinstance(byte_r, six.binary_type):
        byte_r = byte_r.encode('utf8')

    if len(byte_l) != len(byte_r):
        return False

    result = reduce(lambda r, b: r | (b[0] ^ b[1]), map(lambda b: (ord(b[0]), ord(b[1])) if six.PY2 else b, zip(byte_l, byte_r)), 0)

    return result == 0


def generate_message(required_headers, headers, host=None, method=None, path=None, http_version=None):
    headers = CaseInsensitiveDict(headers)

    if not required_headers:
        required_headers = ['date']

    signable_list = []
    for header in required_headers:
        header = header.lower()
        if header == '(request-target)':  # draft-03 to draft-07
            if not method or not path:
                raise KeyError('method and path arguments required when using "(request-target)"')
            signable_list.append('%s: %s %s' % (header, method.lower(), path))
        elif header == '(request-line)':  # draft-02
            if not method or not path:
                raise KeyError('method and path arguments required when using "(request-line)"')
            signable_list.append('%s: %s %s' % (header, method.lower(), path))
        elif header == 'request-line':   # draft-00, draft-01
            if not method or not path or not http_version:
                raise KeyError('method, path and http_version arguments required when using "request-line"')
            signable_list.append('%s %s %s' % (method, path, http_version))

        elif header == 'host':
            # 'host' special case due to requests lib restrictions
            # 'host' is not available when adding auth so must use a param
            # if no param used, defaults back to the 'host' header
            if not host:
                if 'host' in headers:
                    host = headers[header]
                else:
                    raise KeyError('missing required header "%s"' % (header))
            signable_list.append('%s: %s' % (header, host))
        else:
            if header not in headers:
                raise KeyError('missing required header "%s"' % (header))

            signable_list.append('%s: %s' % (header, headers[header]))

    signable = '\n'.join(signable_list).encode('ascii')
    return signable


def parse_authorization_header(header):
    if not isinstance(header, six.string_types):
        header = header.decode('ascii')  # HTTP headers cannot be Unicode.

    auth = header.split(' ', 1)
    if len(auth) < 2:
        raise ValueError('Invalid authorization header. (eg. Method key1=value1,key2="value, \"2\"")')

    # Split up any args into a dictionary.
    values = {}
    auth_value = auth[1]

    # This is tricky string magic.  Let urllib do it.
    fields = parse_http_list(auth_value)
    for item in fields:
        # Only include keypairs.
        if '=' in item:
            # Split on the first '=' only.
            key, value = item.split('=', 1)
            if not (len(key) and len(value)):
                continue

            # Unquote values, if quoted.
            if value[0] == '"':
                value = value[1:-1]

            values[key] = value

    # ("Signature", {"headers": "date", "algorithm": "hmac-sha256", ... })
    return (auth[0], CaseInsensitiveDict(values))


def build_signature_template(key_id, algorithm, headers):
    """
    Build the Signature template for use with the Authorization header.

    key_id is the mandatory label indicating to the server which secret to use
    algorithm is one of the six specified algorithms
    headers is a list of http headers to be included in the signing string.

    The signature must be interpolated into the template to get the final Authorization header value.
    """
    param_map = {'keyId': key_id,
                 'algorithm': algorithm,
                 'signature': '%s'}
    if headers:
        headers = [h.lower() for h in headers]
        param_map['headers'] = ' '.join(headers)
    kv_pairs = map('{0[0]}="{0[1]}"'.format, param_map.items())
    kv_string = ','.join(kv_pairs)
    sig_string = 'Signature {0}'.format(kv_string)
    return sig_string


# based on http://stackoverflow.com/a/2082169/151401
class CaseInsensitiveDict(dict):
    def __init__(self, d=None, **kwargs):
        super(CaseInsensitiveDict, self).__init__(**kwargs)
        if d:
            self.update((k.lower(), v) for k, v in six.iteritems(d))

    def __setitem__(self, key, value):
        super(CaseInsensitiveDict, self).__setitem__(key.lower(), value)

    def __getitem__(self, key):
        return super(CaseInsensitiveDict, self).__getitem__(key.lower())

    def __contains__(self, key):
        return super(CaseInsensitiveDict, self).__contains__(key.lower())


# currently busted...
def get_fingerprint(key):
    """
    Takes an ssh public key and generates the fingerprint.

    See: http://tools.ietf.org/html/rfc4716 for more info
    """
    if key.startswith('ssh-rsa'):
        key = key.split(' ')[1]
    else:
        regex = r'\-{4,5}[\w|| ]+\-{4,5}'
        key = re.split(regex, key)[1]

    key = key.replace('\n', '')
    key = key.strip().encode('ascii')
    key = base64.b64decode(key)
    fp_plain = hashlib.md5(key).hexdigest()
    return ':'.join(a + b for a, b in zip(fp_plain[::2], fp_plain[1::2]))
